### PR TITLE
APEXMALHAR-2151 delimited file format support to FSLoader

### PR DESCRIPTION
**Pull request by shubham-pathak22**
@chinmaykolhatkar Could you please review

### CODE REVIEW

File: contrib/src/main/java/com/datatorrent/contrib/enrich/JsonFSLoader.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import org.codehaus.jackson.type.TypeReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This implementation of {@link FSLoader} is used to load data from json file.
+ * <p>
+ * The format of the input file is:
+ * <p>
+ * {"productCategory": 5, "productId": 0} {"productCategory": 4, "productId": 1}
```
Comment (chinmaykolhatkar) [inline contrib/src/main/java/com/datatorrent/contrib/enrich/JsonFSLoader.java]: One JSON on one line.

**Comment by shubham-pathak22** [comment]
Incorporated review comments

File: contrib/src/main/java/com/datatorrent/contrib/enrich/FSLoader.java [hunk 2]
```diff
@@ -128,6 +112,16 @@ public void setFileName(String fileName)
     return result;
   }
 
+  /**
+   * Extracts values from a record. Concrete implementations override this
+   * method to parse a record and convert it to Map of field names and values
```
Comment (chinmaykolhatkar) [inline contrib/src/main/java/com/datatorrent/contrib/enrich/FSLoader.java]: Can you add in the comment what happens when null is returned? I see both the implementations of this method returns null.

File: contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+   * 
+   * @param fieldDescription
+   *          String representing fields and its type in the same order as it
+   *          appears in the records. E.g.
+   *          empId:INTEGER,address:STRING,phone:LONG Supported data types are
+   *          BOOLEAN, INTEGER, FLOAT, DOUBLE, LONG, SHORT, CHARACTER, STRING
+   */
+  public DelimitedFSLoader(String fieldDescription)
+  {
+    fields = new ArrayList<Field>();
```
Comment (chinmaykolhatkar) [inline contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java]: No need to mention Field as type in ArrayList.

File: contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    List<String> headers = new ArrayList<String>();
```
Comment (chinmaykolhatkar) [inline contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java]: String in type of ArrayList is not required.

File: contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+   * 
+   * @param fieldDescription
+   *          String representing fields and its type in the same order as it
+   *          appears in the records. E.g.
+   *          empId:INTEGER,address:STRING,phone:LONG Supported data types are
+   *          BOOLEAN, INTEGER, FLOAT, DOUBLE, LONG, SHORT, CHARACTER, STRING If
+   *          header is present, field names must match the header.
+   * @param hasHeader
+   *          If set to true, header will be skipped
+   * @param delimiter
```
Comment (chinmaykolhatkar) [inline contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java]: Description missing for delimiter param.

File: contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+  /**
+   * 
+   * @param fieldDescription
+   *          String representing fields and its type in the same order as it
+   *          appears in the records. E.g.
+   *          empId:INTEGER,address:STRING,phone:LONG Supported data types are
+   *          BOOLEAN, INTEGER, FLOAT, DOUBLE, LONG, SHORT, CHARACTER, STRING
+   * @param hasHeader
+   *          If set to true, the first line will be skipped
+   * @param delimiter
```
Comment (chinmaykolhatkar) [inline contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java]: Description missing for delimiter and quote param

File: contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+  /**
+   * 
+   * @param fieldDescription
+   *          String representing fields and its type in the same order as it
+   *          appears in the records. E.g.
+   *          empId:INTEGER,address:STRING,phone:LONG Supported data types are
+   *          BOOLEAN, INTEGER, FLOAT, DOUBLE, LONG, SHORT, CHARACTER, STRING
+   */
+  public DelimitedFSLoader(String fieldDescription)
```
Comment (chinmaykolhatkar) [inline contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java]: You need to have empty constructor for this and let user set the properties for DelimitedFSLoader using properties file.
Have setter method for individual parameters for this FSLoader.

Comment (chinmaykolhatkar) [inline contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java]: Constructor is not necessary.

File: contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+      case FLOAT:
+        return Float.parseFloat(value);
+      case LONG:
+        return Long.parseLong(value);
+      case SHORT:
+        return Short.parseShort(value);
+      case CHARACTER:
+        return value.charAt(0);
+      default:
+        return value;
```
Comment (chinmaykolhatkar) [inline contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java]: What about DATE FIELD_TYPE? I think you need to parse that too. Also which date format is allowed?

**Comment by shubham-pathak22** [comment]
@chinmaykolhatkar  I have used Delimited Schema used in CsvParser to get field-specific information . Updated tests too.

File: contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+import com.datatorrent.contrib.parser.DelimitedSchema.Field;
+import com.datatorrent.lib.util.ReusableStringReader;
+
+/**
+ * This implementation of {@link FSLoader} is used to load data from delimited
+ * file.User needs to provide a schema as a string specified in a json format as
+ * per {@link DelimitedSchema} that contains information of name and type of
+ * field
+ */
+public class DelimitedFSLoader extends FSLoader
```
Comment (chinmaykolhatkar) [inline contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java:48]: Please mark this class as Evolving.

File: contrib/src/main/java/com/datatorrent/contrib/enrich/JsonFSLoader.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+ * {"productCategory": 5, "productId": 2}
+ * {"productCategory": 5, "productId": 3}
+ * </p>
+ * Each line in the input file should be a valid json object which represents a
+ * record and each key/value pair in that json object represents the
+ * fields/value.
+ * <p>
+ *
+ */
+public class JsonFSLoader extends FSLoader
```
Comment (chinmaykolhatkar) [inline contrib/src/main/java/com/datatorrent/contrib/enrich/JsonFSLoader.java:49]: Please mark this class as Evolving.

Comment (shubham-pathak22) [inline contrib/src/main/java/com/datatorrent/contrib/enrich/JsonFSLoader.java:49]: Done

File: contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java (hunk 9 above)
Comment (shubham-pathak22) [inline contrib/src/main/java/com/datatorrent/contrib/enrich/DelimitedFSLoader.java:48]: Done

**Comment by chinmaykolhatkar** [comment]
Looks good to me. I'll merge it in a day if there are no further comments.

**Pull request merged**